common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -365,12 +368,12 @@ public boolean equals(Object other) {
     if (other == null || !(other instanceof CalendarInterval)) return false;
 
     CalendarInterval o = (CalendarInterval) other;
-    return this.months == o.months && this.microseconds == o.microseconds;
+    return this.months == o.months && this.days == o.days && this.microseconds == o.microseconds;
   }
 
   @Override
   public int hashCode() {
-    return 31 * months + (int) microseconds;
+    return 31 * (31 * months + days) + (int) microseconds;
```

Comment:
Not related to the PR but I am wondering why the most significant bits of microseconds are completely excluded from the hash.